tools/ci_build/reduce_op_kernels.py

```diff
@@ -0,0 +1,256 @@
+
+log = get_logger("reduce_op_kernels")
+
+
+# valid C++ scalar types that can be specified as globally allowed types
+_valid_allowed_types = {
+    "bool",
+    "uint8_t", "uint16_t", "uint32_t", "uint64_t",
+    "int8_t", "int16_t", "int32_t", "int64_t",
+    "MLFloat16", "BFloat16",  # in onnxruntime namespace
+    "float", "double",
+    "string",  # in std namespace
+}
+
+
+def _validated_globally_allowed_types(globally_allowed_types: typing.Collection[str]) -> typing.Set[str]:
+    '''Return a valid set of globally allowed types.'''
+    # ensure globally_allowed_types is a set
+    if not isinstance(globally_allowed_types, set):
+        globally_allowed_types = set(globally_allowed_types)
+
+    if not globally_allowed_types <= _valid_allowed_types:
+        raise ValueError(
+            "Globally allowed types must be a subset of valid allowed types. Actual: {}, valid: {}".format(
+                globally_allowed_types, sorted(_valid_allowed_types)))
+
+    return globally_allowed_types
+
+
+def _type_re_from_globally_allowed_types(globally_allowed_types: typing.Set[str]) -> typing.re.Pattern:
+    '''Return a regular expression to match type registration strings to a set of globally allowed types.'''
+    # to keep a registration, the type should match patterns like:
+    # 1. T0
+    # 2. T0_T1_T2
+    # where Ti is a member of globally_allowed_types and multiple Ti's are delimited by "_"
+    # this covers both the common case (1) and special cases like OneHot registration (2)
+    allowed_type_subpattern = \
+        "(?:" + "|".join(re.escape(allowed_type) for allowed_type in sorted(globally_allowed_types)) + ")"
+    return re.compile("^{0}(?:_{0})*$".format(allowed_type_subpattern))
+
+
+class _ExcludingRegistrationProcessor(op_registration_utils.RegistrationProcessor):
+    '''Registration processor that excludes registrations and writes the result to an output file.'''
+    def __init__(self, required_ops: dict, op_type_usage_manager: typing.Optional[OperatorTypeUsageManager],
+                 globally_allowed_types: typing.Optional[typing.Set[str]], output_file: str):
+        self._required_ops = required_ops
+
+        if op_type_usage_manager is not None and globally_allowed_types is not None:
+            raise ValueError("At most one of op_type_usage_manager and globally_allowed_types may be provided.")
+
+        self._op_type_usage_manager = op_type_usage_manager
+
+        self._globally_allowed_types_re = \
+            _type_re_from_globally_allowed_types(globally_allowed_types) \
+            if globally_allowed_types is not None else None
+
+        self._output_file = output_file
+
+    def _is_op_required(self, domain: str, operator: str,
+                        start_version: int, end_version: typing.Optional[int]) -> typing.Tuple[bool, str]:
+        '''See if an op should be excluded because it is not required.'''
+        if domain not in self._required_ops:
```

Comment:
We need to support just reducing on types, so may need to insert something like this here

'''
        self._enable_all_ops = globally_allowed_types is not None and not required_ops

    def _is_op_required(self, domain: str, operator: str,
                        start_version: int, end_version: typing.Optional[int]) -> typing.Tuple[bool, str]:
        '''See if an op should be excluded because it is not required.'''
        if self._enable_all_ops:
            return True
'''